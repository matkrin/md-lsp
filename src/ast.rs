use markdown::{
    mdast::{FootnoteReference, Heading, Link, LinkReference, Node, Text},
    unist::{Point, Position},
};
use regex::Regex;

struct Extracted {
    content: String,
    start_position: usize,
    line_number: usize,
}

impl Extracted {
    fn link_text_node(&self, start_line: usize) -> Node {
        let link_text = Text {
            value: self.content.clone(),
            position: Some(Position {
                start: Point {
                    line: start_line + self.line_number,
                    column: self.start_position,
                    offset: self.start_position,
                },
                end: Point {
                    line: start_line + self.line_number,
                    column: self.start_position + self.content.len(),
                    offset: self.start_position + self.content.len(),
                },
            }),
        };
        Node::Text(link_text)
    }

    fn link_node(&self, start_line: usize) -> Node {
        let link_text_node = self.link_text_node(start_line);

        let link = Link {
            children: vec![link_text_node],
            position: Some(Position {
                start: Point {
                    line: start_line + self.line_number,
                    column: self.start_position - 2,
                    offset: self.start_position - 2,
                },
                end: Point {
                    line: start_line + self.line_number,
                    column: self.start_position + self.content.len() + 2,
                    offset: self.start_position + self.content.len() + 2,
                },
            }),
            url: self.content.clone(),
            title: Some(self.content.clone()),
        };
        Node::Link(link)
    }
}

fn extract_links(input: &str) -> Vec<Extracted> {
    let re = Regex::new(r"\[\[([\s\S]*?)\]\]").unwrap();

    input
        .lines()
        .enumerate()
        .flat_map(|(line_number, line)| {
            re.captures_iter(line).filter_map(move |captures| {
                captures.get(1).map(|content| Extracted {
                    content: content.as_str().to_string(),
                    start_position: content.start() + 1,
                    line_number,
                })
            })
        })
        .collect()
}

pub fn parse_wiki_links(node: &mut Node) {
    let mut links = Vec::new();
    if let Some(children) = node.children() {
        for child in children {
            if let Node::Text(t) = child {
                if t.value.contains("[[") && t.value.contains("]]") {
                    let t_position = t.position.as_ref().unwrap().clone();
                    let extracted = extract_links(&t.value);

                    for i in extracted {
                        let link_ast = i.link_node(t_position.start.line);
                        links.push(link_ast);
                    }
                }
            }
        }
    }
    if let Some(children) = node.children_mut() {
        children.append(&mut links);
    }
    // recurse through children
    if let Some(children) = node.children_mut() {
        for child in children {
            parse_wiki_links(child);
        }
    }
}

pub fn find_heading_for_url<'a>(node: &'a Node, link_url: &str) -> Option<&'a Heading> {
    if let Node::Heading(heading) = node {
        if let Some(Node::Text(Text { value, .. })) = heading.children.first() {
            if value == &link_url.replace('#', "") {
                return Some(heading);
            }
        }
    };

    // recurse through children
    if let Some(children) = node.children() {
        for child in children {
            if let Some(n) = find_heading_for_url(child, link_url) {
                return Some(n);
            }
        }
    }
    None
}

pub fn find_link_for_position(node: &Node, line: u32, character: u32) -> Option<&Node> {
    match node {
        Node::Link(Link { position, .. })
        | Node::LinkReference(LinkReference { position, .. })
        | Node::FootnoteReference(FootnoteReference { position, .. }) => {
            if let Some(pos) = position {
                if (line + 1) as usize >= pos.start.line
                    && (line + 1) as usize <= pos.end.line
                    && ((character + 1) as usize) >= pos.start.column
                    && ((character + 1) as usize) <= pos.end.column
                {
                    return Some(node);
                }
            }
        }
        _ => {}
    };

    // recurse through children
    if let Some(children) = node.children() {
        for child in children {
            if let Some(n) = find_link_for_position(child, line, character) {
                return Some(n);
            }
        }
    }
    None
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_make_wiki_links() {
        let md = "Wikilink [[link]] in paragraph";
        let mut ast = markdown::to_mdast(md, &markdown::ParseOptions::gfm()).unwrap();
        // dbg!(&ast);
        parse_wiki_links(&mut ast);
        // dbg!(&ast);
    }

    #[test]
    fn test_make_wiki_links2() {
        let md = "Another line with [[link1]]
Wikilink [[link]] in paragraph ";
        let mut ast = markdown::to_mdast(md, &markdown::ParseOptions::gfm()).unwrap();
        dbg!(&ast);
        parse_wiki_links(&mut ast);
        dbg!(&ast);
    }
}
