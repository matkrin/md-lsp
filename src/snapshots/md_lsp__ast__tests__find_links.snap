---
source: src/ast.rs
expression: links
---
[
    Link {
        children: [
            Text {
                value: "Some Markdown features (github flavored)",
                position: Some(
                    4:4-4:44 (64-104),
                ),
            },
        ],
        position: Some(
            4:3-4:88 (63-148),
        ),
        url: "#some-markdown-features-(github-flavored)",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Heading 1",
                position: Some(
                    5:4-5:13 (152-161),
                ),
            },
        ],
        position: Some(
            5:3-5:26 (151-174),
        ),
        url: "#heading-1",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Heading 2",
                position: Some(
                    6:6-6:15 (180-189),
                ),
            },
        ],
        position: Some(
            6:5-6:28 (179-202),
        ),
        url: "#heading-2",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Heading 3",
                position: Some(
                    7:8-7:17 (210-219),
                ),
            },
        ],
        position: Some(
            7:7-7:30 (209-232),
        ),
        url: "#heading-3",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Subscript",
                position: Some(
                    8:8-8:17 (240-249),
                ),
            },
        ],
        position: Some(
            8:7-8:30 (239-262),
        ),
        url: "#subscript",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Superscript",
                position: Some(
                    9:8-9:19 (270-281),
                ),
            },
        ],
        position: Some(
            9:7-9:34 (269-296),
        ),
        url: "#superscript",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Lists",
                position: Some(
                    10:8-10:13 (304-309),
                ),
            },
        ],
        position: Some(
            10:7-10:22 (303-318),
        ),
        url: "#lists",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Unordered",
                position: Some(
                    11:10-11:19 (328-337),
                ),
            },
        ],
        position: Some(
            11:9-11:32 (327-350),
        ),
        url: "#unordered",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Ordered",
                position: Some(
                    12:10-12:17 (360-367),
                ),
            },
        ],
        position: Some(
            12:9-12:28 (359-378),
        ),
        url: "#ordered",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Nested",
                position: Some(
                    13:10-13:16 (388-394),
                ),
            },
        ],
        position: Some(
            13:9-13:26 (387-404),
        ),
        url: "#nested",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Task List",
                position: Some(
                    14:8-14:17 (412-421),
                ),
            },
        ],
        position: Some(
            14:7-14:30 (411-434),
        ),
        url: "#task-list",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Code",
                position: Some(
                    15:8-15:12 (442-446),
                ),
            },
        ],
        position: Some(
            15:7-15:20 (441-454),
        ),
        url: "#code",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Inline",
                position: Some(
                    16:10-16:16 (464-470),
                ),
            },
        ],
        position: Some(
            16:9-16:26 (463-480),
        ),
        url: "#inline",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Block",
                position: Some(
                    17:10-17:15 (490-495),
                ),
            },
        ],
        position: Some(
            17:9-17:24 (489-504),
        ),
        url: "#block",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Quotes",
                position: Some(
                    18:8-18:14 (512-518),
                ),
            },
        ],
        position: Some(
            18:7-18:24 (511-528),
        ),
        url: "#quotes",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Links",
                position: Some(
                    19:8-19:13 (536-541),
                ),
            },
        ],
        position: Some(
            19:7-19:22 (535-550),
        ),
        url: "#links",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Images",
                position: Some(
                    20:8-20:14 (558-564),
                ),
            },
        ],
        position: Some(
            20:7-20:24 (557-574),
        ),
        url: "#images",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Footnotes",
                position: Some(
                    21:8-21:17 (582-591),
                ),
            },
        ],
        position: Some(
            21:7-21:30 (581-604),
        ),
        url: "#footnotes",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Tables",
                position: Some(
                    22:8-22:14 (612-618),
                ),
            },
        ],
        position: Some(
            22:7-22:24 (611-628),
        ),
        url: "#tables",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Colors",
                position: Some(
                    23:8-23:14 (636-642),
                ),
            },
        ],
        position: Some(
            23:7-23:24 (635-652),
        ),
        url: "#colors",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Alerts",
                position: Some(
                    24:8-24:14 (660-666),
                ),
            },
        ],
        position: Some(
            24:7-24:24 (659-676),
        ),
        url: "#alerts",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "Comments",
                position: Some(
                    25:8-25:16 (684-692),
                ),
            },
        ],
        position: Some(
            25:7-25:28 (683-704),
        ),
        url: "#comments",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "LinkRefernce and Definition",
                position: Some(
                    26:8-26:35 (712-739),
                ),
            },
        ],
        position: Some(
            26:7-26:66 (711-770),
        ),
        url: "#linkrefernce-and-definition",
        title: None,
    },
    Link {
        children: [
            Text {
                value: "here",
                position: Some(
                    96:31-96:35 (1466-1470),
                ),
            },
        ],
        position: Some(
            96:30-96:71 (1465-1506),
        ),
        url: "https://github.com/matkrin/md-lsp",
        title: None,
    },
]
